File: rust/src/bin/start_agent.rs
//! start-agent CLI command
//! Start an agent with specified configuration

use agent_commander::{
    agent, parse_start_agent_args, show_start_agent_help, validate_start_agent_options,
    AgentOptions, AgentStartOptions,
};

fn parse_tool_env(entries: Vec<String>) -> Result<Vec<(String, String)>, String> {
    entries
        .into_iter()
        .map(|entry| {
            entry
                .split_once('=')
                .map(|(key, value)| (key.to_string(), value.to_string()))
                .ok_or_else(|| format!("Invalid --tool-env value \"{}\". Use KEY=VALUE.", entry))
        })
        .collect()
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let options = parse_start_agent_args(&args);

    // Show help if requested
    if options.help {
        show_start_agent_help();
        std::process::exit(0);
    }

    // Validate options
    let validation = validate_start_agent_options(&options);
    if !validation.valid {
        eprintln!("Error: Invalid options\n");
        for error in &validation.errors {
            eprintln!("  - {}", error);
        }
        eprintln!("\nRun \"start-agent --help\" for usage information.");
        std::process::exit(1);
    }

    let extra_env = match parse_tool_env(options.tool_env) {
        Ok(extra_env) => extra_env,
        Err(error) => {
            eprintln!("Error: {}", error);
            std::process::exit(1);
        }
    };

    // Create agent controller
    let agent_options = AgentOptions {
        tool: options.tool.unwrap_or_default(),
        working_directory: options.working_directory.unwrap_or_default(),
        prompt: options.prompt,
        prompt_file: options.prompt_file,
        system_prompt: options.system_prompt,
        append_system_prompt: options.append_system_prompt,
        model: options.model,
        fallback_model: options.fallback_model,
        resume: options.resume,
        verbose: options.verbose,
        replay_user_messages: options.replay_user_messages,
        session_id: options.session_id,
        fork_session: options.fork_session,
        read_only: options.read_only,
        plan_only: options.plan_only,
        executable: options.tool_executable,
        extra_args: options.tool_args,
        extra_env,
        skip_default_safety_flags: options.skip_default_safety_flags,
        isolation: options.isolation,
        screen_name: options.screen_name,
        container_name: options.container_name,
        ..Default::default()
    };

    let mut controller = match agent(agent_options) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
    };

    // Start the agent
    let start_options = AgentStartOptions {
        dry_run: options.dry_run,
        detached: options.detached,
        attached: options.attached,
    };

    if let Err(e) = controller.start(start_options).await {
        eprintln!("Error: {}", e);
        std::process::exit(1);
    }

    // If not detached, wait for completion
    if !options.detached && !options.dry_run {
        match controller.stop(Default::default()).await {
            Ok(result) => {
                std::process::exit(result.exit_code);
            }
            Err(e) => {
                eprintln!("Error: {}", e);
                std::process::exit(1);
            }
        }
    }
}

File: js/bin/start-agent.mjs
#!/usr/bin/env node

/**
 * start-agent CLI command
 * Start an agent with specified configuration
 */

import { agent } from '../src/index.mjs';
import {
  parseStartAgentArgs,
  showStartAgentHelp,
  validateStartAgentOptions,
} from '../src/cli-parser.mjs';
import {
  parseNdjsonLine,
  stringifyNdjsonLine,
} from '../src/streaming/ndjson.mjs';

/**
 * Build a per-command approval consumer bridge over the process's own stdio.
 *
 * The agent controller normalizes each native permission request and hands it
 * to `onPermissionRequest`. Here we emit that normalized request to stdout as
 * an NDJSON line and wait for the consumer to reply on stdin with a matching
 * `{ "type": "permission_response", "id": <id>, "decision": "once|always|reject" }`
 * frame. The decision is resolved back to the controller, which forwards it to
 * the native CLI's stdin in the correct wire format.
 *
 * @returns {Function} async (normalizedRequest) => 'once' | 'always' | 'reject'
 */
function createStdioApprovalBridge() {
  const pending = new Map();
  let buffer = '';

  process.stdin.setEncoding('utf8');
  process.stdin.on('data', (chunk) => {
    buffer += chunk;
    const lines = buffer.split('\n');
    buffer = lines.pop() || '';
    for (const line of lines) {
      const message = parseNdjsonLine({ line });
      if (
        message &&
        message.type === 'permission_response' &&
        message.id !== null &&
        message.id !== undefined &&
        pending.has(message.id)
      ) {
        const resolve = pending.get(message.id);
        pending.delete(message.id);
        resolve(message.decision);
      }
    }
  });

  return (request) =>
    new Promise((resolve) => {
      pending.set(request.id, resolve);
      // `raw` carries the unnormalized native frame; omit it from the relayed
      // event so the consumer sees a clean, tool-agnostic request.
      const { raw: _raw, ...relayed } = request;
      process.stdout.write(stringifyNdjsonLine({ value: relayed }));
    });
}

async function main() {
  const args = process.argv.slice(2);
  const options = parseStartAgentArgs(args);

  // Show help if requested
  if (options.help) {
    showStartAgentHelp();
    process.exit(0);
  }

  // Validate options
  const validation = validateStartAgentOptions(options);
  if (!validation.valid) {
    console.error('Error: Invalid options\n');
    validation.errors.forEach((error) => console.error(`  - ${error}`));
    console.error('\nRun "start-agent --help" for usage information.');
    process.exit(1);
  }

  try {
    // When per-command approval is requested, bridge the relay to this
    // process's stdio so an external consumer can approve/reject each command.
    const onPermissionRequest = options.approveEach
      ? createStdioApprovalBridge()
      : undefined;

    // Create agent controller
    const agentController = agent({
      tool: options.tool,
      workingDirectory: options.workingDirectory,
      prompt: options.prompt,
      promptFile: options.promptFile,
      systemPrompt: options.systemPrompt,
      model: options.model,
      resume: options.resume,
      readOnly: options.readOnly,
      planOnly: options.planOnly,
      approveEach: options.approveEach,
      onPermissionRequest,
      isolation: options.isolation,
      screenName: options.screenName,
      containerName: options.containerName,
      toolOptions: {
        appendSystemPrompt: options.appendSystemPrompt,
        fallbackModel: options.fallbackModel,
        verbose: options.verbose,
        replayUserMessages: options.replayUserMessages,
        sessionId: options.sessionId,
        forkSession: options.forkSession,
        executable: options.toolExecutable,
        extraArgs: options.toolArgs,
        extraEnv: options.toolEnv,
        skipDefaultSafetyFlags: options.skipDefaultSafetyFlags,
      },
    });

    // Start the agent
    await agentController.start({
      dryRun: options.dryRun,
      detached: options.detached,
      attached: options.attached,
    });

    // Exit with the agent's exit code
    if (!options.detached && !options.dryRun) {
      const result = await agentController.stop();
      process.exit(result.exitCode);
    }
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

main();
